### ISSUE TITLE
How do RNGs work in torchax

### ISSUE THREAD

**Issue by carlesoctav** [google/torchax#17]
I’ve read the code, and it seems that we just let PyTorch manage the RNGs (I’ve tested it, and it works on eager mode), but I have a question.

I’m not entirely sure how this works, but in JAX we sometimes provide RNG keys (e.g., jax.random.key) as arguments to the loss function. Since these keys are JAX arrays, they’re tracer objects. When we generate random numbers (random boolean, for dropout mask)based on these keys, the results also become tracer objects, which are then captured in the computation graph. Shouldn’t there be an equivalent mechanism for this in the torchax?

**Comment by qihqi** [comment]
Hi,

The current situation is a little bit underdocumented. Let me describe your issue to make sure I got it correctly:
So currently if I have:

```
def rand_plus_one():
  return torch.rand((2,2), device='jax') + 1

print(rand_plus_one())
print(rand_plus_one())
```

The number changes because everytime we generate a random number, we actually use torch.randint to get an random integer, and use that as the JAX RNG key.

However, if I do


```
@torchax.interop.jax_jit
def rand_plus_one():
  return torch.rand((2,2), device='jax') + 1

print(rand_plus_one())
print(rand_plus_one())
```

Then, you get same number twice. Because a jax.jit'd function is assumed to be a pure function. We can inspect the StableHLO graph and see that it is burning in the RNG:


```
def rand_plus_one():
  return torch.rand((2,2), device='jax') + 1

print(jax.jit(torchax.interop.jax_view(rand_plus_one)).lower().as_text())
```

```
  func.func public @main() -> (tensor<2x2xf32> {jax.result_info = "result"}) {
    %c = stablehlo.constant dense<[337939533, 4234946772]> : tensor<2xui32> <------ HERE burned in constant
    %0 = call @_threefry_split(%c) : (tensor<2xui32>) -> tensor<2x2xui32>
    %1 = stablehlo.slice %0 [1:2, 0:2] : (tensor<2x2xui32>) -> tensor<1x2xui32>
    %2 = stablehlo.reshape %1 : (tensor<1x2xui32>) -> tensor<2xui32>
    %cst = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %cst_0 = stablehlo.constant dense<1.000000e+00> : tensor<f32>
    %3 = call @_uniform(%2, %cst, %cst_0) : (tensor<2xui32>, tensor<f32>, tensor<f32>) -> tensor<2x2xf32>
    %cst_1 = stablehlo.constant dense<1.000000e+00> : tensor<f32>
    %4 = stablehlo.broadcast_in_dim %cst_1, dims = [] : (tensor<f32>) -> tensor<2x2xf32>
    %5 = stablehlo.add %3, %4 : tensor<2x2xf32>
    return %5 : tensor<2x2xf32>
  }
... more stuff
```

You can set the current RNG with the `manual_seed` function of the `Environment` object. So if I want to create a function that takes RNG as an argument, I can do:

```
env = torchax.default_env()
def rand_plus_one(rng):
  env.manual_seed(rng)
  return torch.rand((2,2), device='jax') + 1

print(jax.jit(torchax.interop.jax_view(rand_plus_one)).lower(2).as_text())
```

Got:


```
...
WARNING:root:Duplicate op registration for aten.__and__
module @jit_call_torch attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<i32>) -> (tensor<2x2xf32> {jax.result_info = "result"}) {
    %c = stablehlo.constant dense<32> : tensor<i32>
    %0 = stablehlo.shift_right_logical %arg0, %c : tensor<i32>
    %1 = stablehlo.convert %0 : (tensor<i32>) -> tensor<ui32>
    %2 = stablehlo.broadcast_in_dim %1, dims = [] : (tensor<ui32>) -> tensor<1xui32>
    %3 = stablehlo.convert %arg0 : (tensor<i32>) -> tensor<ui32>
    %c_0 = stablehlo.constant dense<4294967295> : tensor<ui32>
    %4 = stablehlo.and %3, %c_0 : tensor<ui32>
    %5 = stablehlo.broadcast_in_dim %4, dims = [] : (tensor<ui32>) -> tensor<1xui32>
    %6 = stablehlo.concatenate %2, %5, dim = 0 : (tensor<1xui32>, tensor<1xui32>) -> tensor<2xui32>
    %7 = call @_threefry_split(%6) : (tensor<2xui32>) -> tensor<2x2xui32>
    %8 = stablehlo.slice %7 [1:2, 0:2] : (tensor<2x2xui32>) -> tensor<1x2xui32>
    ...
```

NOTE: while writing this response I noticed a bug so the above only work with https://github.com/google/torchax/pull/20
